### ISSUE TITLE
error when extending a builtin package that uses the python command: "Error: NameError: name 'python' is not defined"

### ISSUE THREAD

**Issue by lukebroskop** [LLNL/spack#27508]
### Steps to reproduce

I attempted to extend the `builtin` petsc package with the following:

```
from spack.pkg.builtin.petsc import Petsc as BuiltinPetsc                                                                                                                                    

import os
from spack import *

class Petsc(BuiltinPetsc):

    def flag_handler(self, name, flags):
        if name == 'ldflags' and self.spec.satisfies('%cce'):
            flags.append("-Wl,-z,notext")
        if name == 'fflags' and self.spec.satisfies('%cce'):
            flags.append("-fPIC")
        return (flags, None, None)
```

@becker33 gave me a solution by adding the following the the extended petsc package:
```
    @run_before('install')
    def fixup_bug(self):
        spack.pkg.builtin.petsc.python = python
```

Greg asked me to open an issue on this since there is a better long-term solution.

### Error message

Here's the error message without adding the fix from Greg. 
```
...
==> Installing petsc-3.15.0-t3jwpmvskoeibjpraoieika3zm6cus57
==> No binary for petsc-3.15.0-t3jwpmvskoeibjpraoieika3zm6cus57 found: installing from source
v==> Using cached archive: /gpfs/alpine/ven114/scratch/lukebr/confidence/spack/var/spack/cache/_source-cache/archive/ac/ac46db6bfcaaec8cd28335231076815bd5438f401a4a05e33736b4f9ff12e59a.tar.gz
==> No patches needed for petsc
==> petsc: Executing phase: 'install'
==> Error: NameError: name 'python' is not defined

/gpfs/alpine/ven114/scratch/lukebr/confidence/spack/var/spack/repos/builtin/packages/petsc/package.py:504, in install:
        501        if '+hpddm' in spec:
        502            options.append('--download-hpddm')
        503
  >>    504        python('configure', '--prefix=%s' % prefix, *options)
        505
        506        # PETSc has its own way of doing parallel make.
        507        make('MAKE_NP=%s' % make_jobs, parallel=False)

See build log for details:
  /tmp/lukebr/spack-stage/spack-stage-petsc-3.15.0-t3jwpmvskoeibjpraoieika3zm6cus57/spack-build-out.txt
...
```

### Information on your system

```
* **Spack:** 0.17.0-170-7c3b146789
* **Python:** 3.8.5
* **Platform:** cray-sles15-zen2
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by jmansour** [comment]
I'm hitting this issue too while trying to extend `numpy`. 

Suggested solution (with appropriate mods) doesn't seem to help unfortunately. 

```
* **Spack:** 0.18.1
* **Python:** 3.6.8
* **Platform:** linux-rhel7-skylake
* **Concretizer:** clingo
```

**Comment by alalazo** [comment]
Fixed in #34157
